packages/language-server/src/core/parseJS.ts

```diff
@@ -94,6 +61,57 @@ export function extractScriptTags(
 	return embeddedJSFiles;
 }
 
+function extractEmbeddedJSFilesRecursively(
+	fileName: string,
+	snapshot: ts.IScriptSnapshot,
+	roots: Node[],
+	array: VirtualFile['embeddedFiles'] = []
+): VirtualFile['embeddedFiles'] {
+	for (const [index, root] of roots.entries()) {
+		if (root.tag !== 'script') continue; // Early return if tag is not 'script'
+
+		if (
+			root.startTagEnd !== undefined &&
+			root.endTagStart !== undefined &&
+			isIsolatedScriptTag(root)
+		) {
+			const scriptText = snapshot.getText(root.startTagEnd, root.endTagStart);
+
+			array.push({
+				fileName: fileName + `.${index}.mts`,
```

Comment:
> Deleted original comment, because I placed it wrong.

I am not quite sure if this is the right approach. Should there be some sort of level visible in the filename? E.g. `Header.astro.0.mts` (root level), but if the root level would have some children `Header.astro.0.0.mts` (first child).

Reply:
The names do need to be unique, but the structure doesn't really matter, feel free to achieve this in any way you see fit!